Content.Server/Radio/EntitySystems/HeadsetSystem.cs

```diff
@@ -86,21 +105,129 @@ private void OnExamined(EntityUid uid, HeadsetComponent component, ExaminedEvent
+    {
+        component.KeyContainer = _container.EnsureContainer<Container>(uid, HeadsetComponent.KeyContainerName);
+    }
 
-        args.PushMarkup(Loc.GetString("examine-headset"));
+    private bool InstallKey(HeadsetComponent src, EntityUid key, EncryptionKeyComponent keyComponent)
+    {
+        if (src.KeyContainer.Insert(key))
+        {
+            UploadChannelsFromKey(src, keyComponent);
+            return true;
+        }
+        return false;
+    }
+    private void UploadChannelsFromKey(HeadsetComponent src, EncryptionKeyComponent key)
+    {
+        foreach (var j in key.Channels)
+            src.Channels.Add(j);
+    }
+    private void RecalculateChannels(HeadsetComponent src)
+    {
+        src.Channels.Clear();
+        foreach (EntityUid i in src.KeyContainer.ContainedEntities)
+        {
+            if (TryComp<EncryptionKeyComponent>(i, out var key))
+            {
+                UploadChannelsFromKey(src, key);
+            }
+        }
+    }
 
-        foreach (var id in component.Channels)
+    private void OnInteractUsing(EntityUid uid, HeadsetComponent component, InteractUsingEvent args)
+    {
+        if (!TryComp<ContainerManagerComponent>(uid, out var storage))
+            return;
+        if(!component.IsKeysLocked)
+        {
+            _popupSystem.PopupEntity(Loc.GetString("headset-encryption-keys-are-locked"), uid, args.User);
+            return;
+        }
+        if (TryComp<EncryptionKeyComponent>(args.Used, out var key))
         {
-            if (id == "Common") continue;
-
-            var proto = _protoManager.Index<RadioChannelPrototype>(id);
-            args.PushMarkup(Loc.GetString("examine-headset-channel",
-                ("color", proto.Color),
-                ("key", proto.KeyCode),
-                ("id", proto.LocalizedName),
-                ("freq", proto.Frequency)));
+            if (component.KeySlots > component.KeyContainer.ContainedEntities.Count)
+            {
+                if (_container.TryRemoveFromContainer(args.Used) && InstallKey(component, args.Used, key))
+                {
+                    _popupSystem.PopupEntity(Loc.GetString("headset-encryption-key-successfully-installed"), uid, args.User);
+                    _audio.PlayPvs(component.KeyInsertionSound, args.Target);
+                }
+            }
+            else
+            {
+                _popupSystem.PopupEntity(Loc.GetString("headset-encryption-key-slots-already-full"), uid, args.User);
+            }
         }
+        if (TryComp<ToolComponent>(args.Used, out var tool))
+        {
+            if (component.KeyContainer.ContainedEntities.Count > 0)
+            {
+                if (_toolSystem.UseTool(
+                    args.Used, args.User, uid,
+                    0f, 0f, new String[] { component.KeysExtractionMethod },
+                    doAfterCompleteEvent: null, toolComponent: tool)
```

Comment:
> Also, the player should have the option to remove a specific key from the headset instead of all of them at once.

This is like in original, but i don't know how to create "input" from list of entities for this.

Reply:
Either use verbs or a UI.